File: ui/scripts/fixtures/codex-retention-matrix.mjs
// Mutates only a matrix-owned Codex database, never the user's shared store.
import assert from "node:assert/strict";
import { DatabaseSync } from "node:sqlite";
import { readdirSync, readFileSync, writeFileSync, existsSync } from "node:fs";
import { resolve, join } from "node:path";

const [directory, retention = "inspect"] = process.argv.slice(2);
const root = resolve(directory);
assert(root.includes("attribution-matrix"), "An isolated matrix directory is mandatory");
const file = readdirSync(root)
  .filter((p) => /^logs_\d+\.sqlite$/.test(p))
  .sort((a, b) => Number(b.match(/\d+/)[0]) - Number(a.match(/\d+/)[0]))[0];
assert(file);
const db = new DatabaseSync(join(root, file));
db.exec("PRAGMA busy_timeout=2000");
const identity =
  process.argv[4] ??
  db
    .prepare(
      "SELECT process_uuid FROM logs WHERE process_uuid LIKE 'pid:%' ORDER BY id DESC LIMIT 1",
    )
    .get()?.process_uuid;
assert(identity);
assert(/^pid:\d+:[a-zA-Z0-9-]+$/.test(identity));
const backup = join(root, `threadless-${identity.replaceAll(":", "-")}.json`);
if (retention !== "inspect") {
  db.exec("BEGIN IMMEDIATE");
  try {
    const rows = db
      .prepare("SELECT * FROM logs WHERE process_uuid=? AND thread_id IS NULL ORDER BY id")
      .all(identity);
    if (!existsSync(backup)) writeFileSync(backup, JSON.stringify(rows));
    const original = JSON.parse(readFileSync(backup, "utf8"));
    assert(original.length);
    const keys = Object.keys(original[0]);
    const restore = db.prepare(
      `INSERT OR IGNORE INTO logs (${keys.join(",")}) VALUES (${keys.map(() => "?").join(",")})`,
    );
    for (const row of original) restore.run(...keys.map((k) => row[k]));
    const insertKeys = keys.filter((k) => k !== "id");
    const late = {
      ...original.at(-1),
      feedback_log_body: "retention matrix late threadless diagnostic",
    };
    const lateId = db
      .prepare(
        `INSERT INTO logs (${insertKeys.join(",")}) VALUES (${insertKeys.map(() => "?").join(",")})`,
      )
      .run(...insertKeys.map((k) => late[k])).lastInsertRowid;
    const predicate = {
      intact: "0",
      initial: `id=${original[0].id}`,
      partial: `id<=${original[Math.floor(original.length / 2)].id}`,
      all: "1",
      late_only: `id<${lateId}`,
    }[retention];
    assert(predicate !== undefined, "Unknown retention case");
    db.prepare(
      `DELETE FROM logs WHERE process_uuid=? AND thread_id IS NULL AND (${predicate})`,
    ).run(identity);
    db.exec("COMMIT");
  } catch (error) {
    db.exec("ROLLBACK");
    throw error;
  }
}
const result = {
  identity,
  file: join(root, file),
  retention,
  counts: db
    .prepare(
      "SELECT COUNT(*) AS total, MIN(id) AS first, MIN(CASE WHEN thread_id IS NULL THEN id END) AS firstThreadless FROM logs WHERE process_uuid=?",
    )
    .get(identity),
  lifecycle: db
    .prepare(
      "SELECT id,thread_id,substr(feedback_log_body,1,2048) AS body FROM logs WHERE process_uuid=? AND (feedback_log_body LIKE 'app_server.request{%rpc.method=\"thread/%' OR feedback_log_body LIKE 'session_loop{%') ORDER BY id",
    )
    .all(identity),
};
console.log(JSON.stringify(result));
db.close();

File: ui/scripts/fixtures/codex-retention-matrix.py
"""Linux counterpart of the native Node SQLite fixture; isolated databases only."""
import json
import re
from pathlib import Path
import sqlite3
import sys

root = Path(sys.argv[1]).resolve()
assert str(root).startswith('/tmp/laymux-attribution-matrix/')
retention = sys.argv[2] if len(sys.argv) > 2 else 'inspect'
path = max(root.glob('logs_*.sqlite'), key=lambda p: int(p.stem.split('_')[1]))
db = sqlite3.connect(path, timeout=2)
db.row_factory = sqlite3.Row
identity = sys.argv[3] if len(sys.argv) > 3 else db.execute("SELECT process_uuid FROM logs WHERE process_uuid LIKE 'pid:%' ORDER BY id DESC LIMIT 1").fetchone()[0]
assert re.fullmatch(r'pid:\d+:[a-zA-Z0-9-]+', identity)
backup = root / ('threadless-' + identity.replace(':', '-') + '.json')
if retention != 'inspect':
    with db:
        db.execute('BEGIN IMMEDIATE')
        rows = [dict(r) for r in db.execute('SELECT * FROM logs WHERE process_uuid=? AND thread_id IS NULL ORDER BY id', (identity,))]
        if not backup.exists():
            backup.write_text(json.dumps(rows))
        original = json.loads(backup.read_text())
        assert original
        keys = list(original[0])
        sql = 'INSERT OR IGNORE INTO logs (' + ','.join(keys) + ') VALUES (' + ','.join('?' for _ in keys) + ')'
        db.executemany(sql, [[row[k] for k in keys] for row in original])
        keys.remove('id')
        late = dict(original[-1], feedback_log_body='retention matrix late threadless diagnostic')
        last = db.execute('INSERT INTO logs (' + ','.join(keys) + ') VALUES (' + ','.join('?' for _ in keys) + ')', [late[k] for k in keys]).lastrowid
        predicate = {'intact': '0', 'initial': f'id={original[0]["id"]}', 'partial': f'id<={original[len(original)//2]["id"]}', 'all': '1', 'late_only': f'id<{last}'}[retention]
        db.execute('DELETE FROM logs WHERE process_uuid=? AND thread_id IS NULL AND (' + predicate + ')', (identity,))
print(json.dumps({'identity': identity, 'file': str(path), 'retention': retention, 'counts': dict(db.execute('SELECT COUNT(*) AS total, MIN(id) AS first, MIN(CASE WHEN thread_id IS NULL THEN id END) AS firstThreadless FROM logs WHERE process_uuid=?', (identity,)).fetchone()), 'lifecycle': [dict(r) for r in db.execute('SELECT id,thread_id,substr(feedback_log_body,1,2048) AS body FROM logs WHERE process_uuid=? AND (feedback_log_body LIKE \'app_server.request{%rpc.method="thread/%\' OR feedback_log_body LIKE \'session_loop{%\') ORDER BY id', (identity,))]}))
